tICA/scripts/ComputeGroupTICA.m

```diff
@@ -0,0 +1,247 @@
+    sICAAvgSpectra = ciftiopen(sICAAvgSpectraName, wbcommand);
+    sICAMaps = ciftiopen(sICAMapsAvgName, wbcommand);
+    sICAVolMaps = ciftiopen(sICAVolMapsAvgName, wbcommand);
+
+    if ~strcmp(tICAMM,'')
+        tICAMM=load(tICAMM);
+    end
+
+    numsubj = length(TCSList);
+    if length(SubjectList) ~= numsubj || length(SpectraList) ~= numsubj
+        error('input lists are not the same length');
+    end
+
+    TCSMask = reshape(TCSMaskConcat.cdata, [sICAdim, RunsXNumTimePoints, numsubj]);
+
+    numfullsubj = 0;
+
+    %tica runwise normalization
+    TCSFullRunVars = single(zeros(sICAdim, RunsXNumTimePoints * numsubj));
+    for i = 1:numsubj
+        if exist(TCSList{i}, 'file')
+            subjBaseInd = (i - 1) * RunsXNumTimePoints + 1;
+            thisStart = subjBaseInd;
+            TCSRunVarSub = [];
+            SubjFolder = [StudyFolder '/' SubjectList{i} '/'];
+            for j = 1:length(fMRIList)
+                dtseriesName = [SubjFolder fMRIList{j}];
+                if exist(dtseriesName, 'file')
+                    [~, runLengthStr] = system(['wb_command -file-information -only-number-of-maps ' dtseriesName]);
+                    runLength = str2double(runLengthStr);
+                    nextStart = thisStart + runLength;
+                    TCSRunVarSub = [TCSRunVarSub repmat(std(TCSFullConcat.cdata(:, thisStart:(nextStart - 1)), [], 2), 1, runLength)];
+
+                    thisStart = nextStart;
+                end
+            end
+            if size(TCSRunVarSub, 2) == RunsXNumTimePoints
+                numfullsubj = numfullsubj + 1;
+            end
+            TCSFullRunVars(:, subjBaseInd - 1 + (1:size(TCSRunVarSub, 2))) = TCSRunVarSub;
+        end
+    end
+    %end tica runwise normalization
+
+    nlfunc = 'tanh';
+    iterations = 100;
+    tICAdim = sICAdim;
+
+    sICAtcsvars = std(TCSFullConcat.cdata, [], 2);
+    TCSFullConcat.cdata = (TCSFullConcat.cdata ./ TCSFullRunVars) .* repmat(sICAtcsvars, 1, size(TCSFullRunVars, 2)); %Making all runs contribute equally improves tICA decompositions
+    TCSFullConcat.cdata(~isfinite(TCSFullConcat.cdata)) = 0;
+
+    %This loop produces more reproducible and better tICA decompositions
+    if strcmp(tICAmode,'ESTIMATE')
+        ITERATIONS=[1:6];
+    elseif strcmp(tICAmode,'INITIALIZE')
+        ITERATIONS=[2:6];
+        A = tICAMM;
+        if size(tICAMM,1) ~= size(TCSFullConcat.cdata,1)
+            error('Initialization mixing matrix does not match dimensions of the sICA components');
+        end
+    elseif strcmp(tICAmode,'USE')
+        ITERATIONS=[0];
+        A = tICAMM;
+        if size(tICAMM,1) ~= size(TCSFullConcat.cdata,1)
+            error('Mixing matrix to be used does not match dimensions of the sICA components');
+        end
+    else
+        error('tICAmode not recognized');
+    end
+    
+    for i = ITERATIONS
```

Comment:
I'm starting to understand the intention behind this.  It would probably be easier to read if, instead of using the loop count as an additional mode switch, to do this, and eliminate the elseif tree that modifies `ITERATIONS`:
```matlab
if strcmp(tICAmode, 'USE')
    <set the "icasso result" to the input file>
else
    if strcmp(tICAmode, 'INITIALIZE')
        <set the initial state to the input file>
    else
        <do icasso, centroid of 100x random initialization>
    end
    <loop, doing icasso iterations>
end
<do "final" step>
```

This should allow the loop to have fixed start and end (2 through 5, though it doesn't look like the loop index gets passed into anything), with no conditionals inside it, if I understand correctly.

Reply:
We do use the loop index in the file names.

Reply:
Oh, I see, there is a whole bunch of extra code after the icasso that runs every iteration...

Reply:
Technically, all the stuff that runs every iteration could be put into a subfunction, so that it could be run in more than one place in the code without bloat, to avoid the weird loop/mode construct, but it probably isn't worth it.  I added a comment about why it is set up that way.